File: Python/A_Project/C_Elevator_System/SchedulingStrategy/ConcreteStrategies/fcfs_scheduling_strategy.py
from CommonEnums.direction import Direction
from SchedulingStrategy.scheduling_strategy import SchedulingStrategy


class FCFSSchedulingStrategy(SchedulingStrategy):
    def get_next_stop(self, elevator) -> int:
        current_floor = elevator.get_current_floor()
        request_queue = elevator.get_request_queue()

        if not request_queue:
            elevator.set_direction(Direction.IDLE)
            return current_floor

        next_request = request_queue[0]
        target_floor = next_request.get_floor()

        if target_floor > current_floor:
            elevator.set_direction(Direction.UP)
        elif target_floor < current_floor:
            elevator.set_direction(Direction.DOWN)
        else:
            elevator.set_direction(Direction.IDLE)

        return target_floor

File: Ruby/A_Project/C_Elevator_System/SchedulingStrategy/ConcreteStrategies/fcfs_scheduling_strategy.rb
require_relative '../scheduling_strategy'
require_relative '../../CommonEnums/direction'

# First-Come-First-Serve scheduling: always serves the oldest request first
class FCFSSchedulingStrategy < SchedulingStrategy
  def get_next_stop(elevator)
    current_floor  = elevator.get_current_floor
    request_queue  = elevator.get_request_queue

    if request_queue.empty?
      elevator.set_direction(Direction::IDLE)
      return current_floor
    end

    target_floor = request_queue.first.get_floor

    if target_floor > current_floor
      elevator.set_direction(Direction::UP)
    elsif target_floor < current_floor
      elevator.set_direction(Direction::DOWN)
    else
      elevator.set_direction(Direction::IDLE)
    end

    target_floor
  end
end
